Eliminate use of DPMI memory when $_dpmi=(0).

bartoldeman opened the pull request:
Up so far we've always had a vbase for virtual hardware, but with extmem/XMS the vbase is only needed for DPMI, whereas the physical RAM always exists.

With the LFB it's similar, except it's convenient to keep the LFB vbase constant; XMS mappings are temporary, and we've already disabled the LFB completely with config.dpmi==0, and that cannot be done with XMS.

Allowing register_hardware_ram without a vbase makes it possible to use a minimal main_pool of just 0x110000 if $_dpmi=(0).
    
My goal is to only alias map and unmap that extmem/XMS virtual hardware ram on demand in DPMI fn 0800/0801 in a followup PR.

With VCPI extmem/XMS will need to be mapped into KVM again, but that should be just physical memory (where we don't know the virtual address unless we inspect the VCPI client's page tables), not something in mem_base with a specific dosaddr associated to it.

bartoldeman committed:
low_mem_init: put memchecks and DPMI init together.

A small cleanup.

bartoldeman commented:
For context, here is a followup patch (not to apply):
https://gist.github.com/cf5ea1505a8137f9b01b0503d9b319de
which allocates and alias maps the DPMI virtual memory on demand in int31/ax=0800, instead of pre-alllocating an area that could alias map all XMS memory in principle.

This passes the XMS test but it cheats a little: in KVM the XMS alias for DPMI is then identity mapped in guest page tables and then mapped to the right SHM region instead of actually using the physical RAM that dosemu reports.

That could be fixed by extending kvm_base, alias mapping from it to SHM, and adjusting guest page tables.

stsp commented on src/base/init/init.c:
```diff
@@ -428,13 +428,13 @@ void low_mem_init(void)
-    result = alias_mapping(MAPPING_EXTMEM, DOSADDR_REL(ptr2),
+  }
+  /* create ext_mem alias for int15/dpmi */
+  result = alias_mapping_pa(MAPPING_EXTMEM, LOWMEM_SIZE + HMASIZE,
```
And suddenly its out of config.dpmi
block?

bartoldeman replied:
yes, this is to associate SHM (`lowmem + LOWMEM_SIZE + HMASIZE`) to int15 memory for `$_dpmi=(0)`.

Without this PR, if `$_dpmi=(0)`, int15 memory is mapped by `physaddr_to_unixaddr` to `&mem_base[0x110000]`,
and else it's mapped to `&lowmem[0x110000]`. With this PR it's always mapped to SHM so it's more consistent.

Should I simply adjust the git commit log entry to mention that, or split the commit?

stsp replied:
Separate commit is always better.

bartoldeman replied:
so I already had the seperate commit (number one in this series).

Just the comment is off. Actually I was confused and I need to change the comment again.

Basically I thought for $_dpmi != (0), int15 was copying to and from SHM (just like xms.c does), whereas in reality it copies in and out of mem_base like you wrote. The purpose of that alias_mapping was to connect that mem_base memory to SHM; with $_dpmi=(0) that SHM memory was allocated (together with lowmem) but not
used at all.

So in this PR I'm changing int15 to copy to and from SHM, but that change is independent of the setting of $_dpmi.

Then alias_mapping_pa sets up a) the connection between phys and SHM (via hwram_update_aliasmap) and b) creates the alias for DPMI if $_dpmi is not 0.

stsp replied:
> creates the alias for DPMI if $_dpmi is not 0.

And yet you move it out of the
config.dpmi check... So where is
the check so that it
"creates the alias for DPMI if $_dpmi is not 0"?

stsp replied:
I still don't understand.
Currently alias_mapping_pa() does this:
`addr2 = mappingdriver->alias(cap, MEM_BASE32(va), mapsize, protect, source);`
If you added some check in the
later patch, then please move it
into this patch. I don't understand
how `virtual mem_base addresses >= 0x110000 are
no longer accessed anywhere` if it
is accessed right in alias_mapping_pa().

stsp replied:
OK I think this all will stop confusing me
if you change the line 426 to use _virtual2(),
so that int15 is referred to lowmem_base,
in the patch before this one.

stsp replied:
So why not to use virtual2 at line
426 and leave that alias_mapping_pa()
as is, i.e. for dpmi only?

stsp replied:
Also reword doesn't help at all:
the patch is still called
`extmem: use alias_mapping_pa to map extmem`
and instead just moves the mapping
out of config.dpmi check.
I am about to ask you to split that
into 2 patches then, if you can't write
a log for this one. :)

bartoldeman replied:
it does change from `alias_mapping` to `alias_mapping_pa` as well, it's not just moving it out of the dpmi check, and `alias_mapping_pa` (unlike `alias_mapping()`) calls `hwram_update_aliasmap()`, which is what is needed for extmem with or without dpmi.

so about the virtual2: if I use:
```
  register_hardware_ram_virtual2('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
            ptr2, DOSADDR_REL(ptr2));
```
then this won't change anything, and the move to kvm_base two patches later won't move it.

On the other hand if I use
```
  register_hardware_ram_virtual2('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
            lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr2));
```
then this is wrong since it includes XMS, not just extmem (the same issue we had before).

So I'm not quite sure what you mean with virtual2. Maybe it's better to split then?
```
  register_hardware_ram_virtual2('X', LOWMEM_SIZE + HMASIZE, EXTMEM_SIZE - HMASIZE,
            lowmem + LOWMEM_SIZE + HMASIZE, DOSADDR_REL(ptr2));
  register_hardware_ram_virtual('x', LOWMEM_SIZE + EXTMEM_SIZE, phys_rsv - (EXTMEM_SIZE - HMASIZE),
            DOSADDR_REL(ptr2) + (EXTMEM_SIZE - HMASIZE));
```
However if I do that, once kvm_base is extended, the extmem part of it isn't used, and there will be fragmentation in the KVM slots, ie.
```
0 -> &kvm_base[0]
0x110000 -> LOWMEM(0x110000) (instead of 0x110000 -> &kvm_base[0x110000] -> LOWMEM(0x110000))
xms_base -> kvm_base[xms_base]
```
so then I need to undo the split again in the 3rd patch?

stsp replied:
> it does change from alias_mapping to alias_mapping_pa as well

Yes, and that part is properly described
in a log, so no questions to that one.

Yes, virtual2 stuff is not needed here, but
I don't understand why we have that
unchangelogged move, so please move
it to another patch then. :)

bartoldeman replied:
it's now split into two patches. Unless you want me to squash patch 2 and patch 4 together?

stsp replied:
No, don't squash, I am trying to understand
this part.
So you say `but no longer accessed by int15()`
which probably applies to previous patch, not
this one. So for this patch I only see:
`This way the extmem mappings are fully consistent with XMS`.
Why not to write what does this mean?
I think that means "extmem now refers the
unix address (used by int15) to lowmem_base
also when dpmi disabled, while the previous
patch made it so only for dpmi enabled".
I surmise that from the prev patch, which has
now the updated description that makes things
clearer.

stsp replied:
But alias_mapping() also updates aliasmap.
So I still don't understand this part then:
```
Because alias_mapping_pa() calls hwram_update_aliasmap(),
physaddr_to_unixaddr() in int15() now maps to an address in
shared memory, instead of in main_pool (mem_base).
```
Why wasn't int15 using shared mem before
if alias_mapping() also calls update_aliasmap()?

bartoldeman replied:
I asked myself the same question and it's because update_aliasmap has this:
```
  if (dosaddr >= ALIAS_SIZE)
    return;

```

bartoldeman replied:
> "extmem now refers the
> unix address (used by int15) to lowmem_base
> also when dpmi disabled, while the previous
> patch made it so only for dpmi enabled".

yes, exactly. I'll try to reword again.

The point of XMS consistency is that it calls `alias_mapping_pa()` in `map_EMB()` always, with or without `config.dpmi`.

stsp replied:
So is this a regression in your prev
patch where you didn't change this
ALIAS_SIZE to eg mem_bases[MEM_BASE].size?
There is definitely some bug here.
Its not possible that you "refuse" to
document some changes or I can't
understand a single-line change for
so long, and no bugs around. :)

bartoldeman replied:
no, it's not a regression. it has to be ALIAS_SIZE there.

in order to use `hwram_update_aliasmap`, `update_aliasmap` would need the physical address from the virtual one (i.e. the reverse of what we do now), and that has never been implemented.

bartoldeman replied:
checking history:
859086eaa had this ok, obviously: `return LOWMEM(addr)`, and 4b970fc20d as well.
d54e82e3cd seems to have changed it.

stsp replied:
> and that has never been implemented.

Right, but seems rather simple:
[a.diff.gz](https://github.com/dosemu2/dosemu2/files/11020840/a.diff.gz)
What do you think?

stsp replied:
Forgot to check addr2 for -1 in the
above patch.

bartoldeman replied:
yes that looks like a good idea! Can you do that as a seperate PR, and then I'll rebase on top of that?

stsp replied:
#1962 
Btw, I was telling you a few months ago
that non-identity mapping would be too
complex, you insisted. :) Well, I didn't
know it'd take us that long though.

stsp commented:
> My goal is to only alias map and unmap that extmem/XMS virtual hardware ram on demand in DPMI
> fn 0800/0801 in a followup PR.

OK, you described the goal, its quite
clear from the patch as well. But I fail
to see a single word on why do we want
that. It was just simpler to always pre-map
xms. Simplicity is a valid goal, as no one
needs to explain why simplicity is needed. :)

bartoldeman commented:
I wanted to separate DPMI specific code (setting up the virtual mapping) from XMS specific code (locking the physical memory). But we can keep premapping XMS, that's indeed simpler.

What I don't like about the present setup is that the physical memory for extmem/XMS in KVM is mapped back into DPMI space. So for that I need to move the physical KVM mapping (GPA->HVA) from
`0x110000 - phys_low -> &mem_base[dpmi_base + dpmi_size + alignment]`
to
`0x110000 - phys_low -> &kvm_base[0x110000]`
(which will be a seperate PR)

once that is done, `&mem_base[dpmi_base + dpmi_size + alignment]` is no longer mapped by KVM so no longer subject to the alignment requirements, which means that then that can then be moved below dpmi_base with all the other linear mappings, so it can be handled exactly the same way as the VGAEMU LFB.

Still I think this PR is then needed to handle the case of `$_dpmi=(0)`, as `&mem_base[0x110000]` and up should not be needed at all if DPMI is disabled.

stsp commented:
But kvm_base is just an alias base, which
ends after HMA. So do you want to enlarge
kvm_base, making it a physical representation?



> Still I think this PR is then needed to handle the case of $_dpmi=(0), as &mem_base[0x110000] and
> up should not be needed at all if DPMI is disabled.

So which patches from this PR are
needed if we keep xms pre-mapped?

bartoldeman commented:
> So which patches from this PR are
> needed if we keep xms pre-mapped?

all of them (to remove the use of linear addresses >= 0x110000, when $_dpmi=(0)), but the comments needed some adjustment.

with xms-premapped the difference will be that DPMI 0800/0801 don't need to be adjusted, since the vbase is registered much like the LFB.

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -1065,14 +1068,16 @@ int alias_mapping_pa(int cap, unsigned addr, size_t mapsize, int protect,
     return 0;
   assert(addr >= LOWMEM_SIZE + HMASIZE);
+  hwram_update_aliasmap(hw, addr, mapsize, source);
+  if (va == (dosaddr_t)-1)
```
So here's the trick, but the patch
that moves a call out of config.dpmi
check goes before?
Looks clearly wrong.
You probably should move that "move"
to this patch.

bartoldeman replied:
I tested the patch before. `alias_mapping_pa` works ok in the first patch much like it does for XMS, because after patch 1 the mem_base area still exists and is registered as vbase for !config.dpmi.

stsp replied:
But the log of the first patch says otherwise:
```
This way with $_dpmi=(0) virtual mem_base addresses >= 0x110000 are
no longer accessed anywhere.
```

bartoldeman replied:
I tried to reword this now. When config.dpmi==0 that memory is still mapped but not accessed. It can't be accessed by regular DOS code since DPMI is disabled and VCPI not merged yet.

The only place that accessed it was `int15()` via `physaddr_to_unixaddr()`.

So after this change that mem_base area is mapped but unused. In the 6th (last) patch in this PR I no longer map it either (which needs the hardware ram with vbase==-1), but in the 3rd patch I already need alias_mapping_pa for extmem to map both virtual and physical memory to shm.

stsp commented on src/base/init/init.c:
```diff
@@ -422,10 +423,14 @@ void low_mem_init(void)
 	    DOSADDR_REL(ptr2));
+  } else
+    /* without DPMI there is only registered physical memory, no alias */
+    register_hardware_ram('X', LOWMEM_SIZE + HMASIZE, phys_rsv);
```
Wasn't that one creating the kmem
maps?

bartoldeman replied:
This used to only register kmem maps, but can be used in general to define any physical region without a virtual address.

The actual kmem mapping happens in init_hardware_ram, and now checks for the type ('h' or 'v')

stsp replied:
OK but it doesn't call mmap_kvm().
So kmems are getting visible to kvm
when mapped to main_pool.
And here?

bartoldeman replied:
here, it's just registered, but never mapped to main_pool, because it doesn't need to be mapped to main_pool any more (for $_dpmi = 0)

kmem won't map either then (because of the explicit config.dpmi check).

stsp replied:
I understand its not mapped to main_pool.
How it is mapped to kvm?

bartoldeman replied:
it's not mapped to kvm any more either, kvm now only maps lowmem+hma and the monitor if dpmi=0. Basically the setting of config.cpu_vm_dpmi is then not relevant any more.

stsp replied:
But isn't that weird?
We use map_hardware_ram() for
things that _are_ mapped to kvm.
And what if vcpi maps that area?

stsp replied:
I mean, why not to use _virtual
as before, just pass -1 as va, and
get things mapped to kvm?

bartoldeman replied:
if you pass -1 to `register_hardware_ram_virtual` it'll call
`  register_hardware_ram_virtual2(type, base, size, MEM_BASE32x(-1, KVM_BASE), -1);`
which makes no sense.

I could extend kvm_base first then, in which case something like:
`  register_hardware_ram_virtual2(type, base, size, &kvm_base[base], -1);`
could work, with appropriate adjustments in virtual2?

stsp replied:
Only for xms I guess?
Because for ext_mem we already have
a physical view at lowmem_base...
Yes, that sounds like a plan.

bartoldeman replied:
done that as a followup patch now. If it's ok I can cleanup the history.

it's easiest to use kvm_base for everything, otherwise KVM will have multiple slots: low -> kvm_base, ext_mem -> lowmem_base, xms -> kvm_base, when alias mapping lowmem_base into kvm_base gives a single mapping for KVM.

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -860,7 +863,8 @@ int register_hardware_ram(int type, unsigned base, unsigned int size)
 	void *uaddr, dosaddr_t va)
 {
-  do_register_hwram(type, base, size, MEM_BASE32(va), va);
+  do_register_hwram(type, base, size,
```
So why still not _virtual2()?

bartoldeman replied:
I am not sure I understand? MEM_BASE32x is a static function, so using _virtual() instead of _virtual2() to map kvm_base.

stsp replied:
Sorry, haven't noticed its mapping.c itself. :)

stsp commented:
> The cost is that now both kvm_base and main_pool need to alias
> map into SHM, for compatibility with native/JIT DPMI.

Not sure I can parse that.
What does this mean, "both kvm_base and main_pool need to alias
map into SHM"?

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -178,8 +178,6 @@ static unsigned char *MEM_BASE32x(dosaddr_t a, int base)
   if (mem_bases[base] == MAP_FAILED)
     return MAP_FAILED;
-  if (base == MEM_BASE || a >= ALIAS_SIZE)
-    return MEM_BASE32(a);
```
Maybe add some assert of alike?
I think its an asking for buffer overrun.

stsp replied:
Actually I think you just need to
enlarge an alias size for kvm_base,
but leave this check alone.
Then you won't need to use virtual2()
explicitly because it will fall-back to
mem_base at the certain address.

bartoldeman replied:
I can certainly add checks to avoid buffer overruns, but in that case I think it's best to then also check mem_base, that is a size for each of the 3 bases. I don't like the fallback, as it caused some confusion in alias_mapping, i.e. it doesn't return MAP_FAILED, so alias_mapping maps the same thing 3 times.

it's also confusing as
`MEM_BASE32x(0x110000, KVM_BASE)` without any change returns `&mem_base[0x110000]` but I would expect it to return `&kvm_base[0x110000]` naturally.

stsp replied:
> it's also confusing as
> MEM_BASE32x(0x110000, KVM_BASE) without any change returns &mem_base[0x110000] but I would 
> expect it to return &kvm_base[0x110000] naturally.

If you enlarge the alias size, then it
will return "&kvm_base[0x110000] naturally",
but not for the high mappings of ext_mem/xms.
I am not forcing that approach, its just to consider.

stsp commented on src/base/init/init.c:
```diff
@@ -411,26 +411,26 @@ void low_mem_init(void)
-    /* register ext_mem and XMS alias memory for dpmi */
-    register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
-	    DOSADDR_REL(ptr2));
+    /* unused hole in physical address space for alignment */
```
You mean entire phys_rsv is now an unused
hole for alignment?

bartoldeman replied:
what I mean here is that the 'U' mapping's physical address is aligned at the next available location above memsize (`=config.dpmi_base + dpmi_mem_size()`). So there's a hole between memsize and the physical location (guest_phys_addr in KVM) that is mapped to `&mem_base[0x110000]`.

stsp commented:
So you don't need the patch that
mangles register_hardware_ram(),
right? It triggers lots of red flags for
me. :)

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -872,7 +876,7 @@ void register_hardware_ram_virtual(int type, unsigned base, unsigned int size,
 {
   register_hardware_ram_virtual2(type, base, size,
-	MEM_BASE32x(va, KVM_BASE), va);
+	MEM_BASE32x(base, KVM_BASE), va);
```
... or possibly you can add the check
here, and fall-back to mem_base.
Then you still don't need to use _virtual2().

bartoldeman replied:
what kind of check do you propose?
`type == 'U' ? MEM_BASE32x(va, MEM_BASE) : MEM_BASE32x(base, KVM_BASE)`
?

stsp replied:
No, check if alias_size have exceeded.

stsp replied:
And if so - fall back to mem_base.

bartoldeman replied:
ok that would be this then?
`base <= ALIAS_SIZE ? MEM_BASE32x(base, KVM_BASE) : MEM_BASE32(va)`

stsp replied:
Maybe it should be something like
`base < mem_bases[KVM_BASE].size ? MEM_BASE32x(base, KVM_BASE) : MEM_BASE32(va)`

bartoldeman replied:
ok. I think I'll just start a new PR just for mem_bases then, as it addresses another issue as well (the triple mapping).

bartoldeman commented:
it indeed reverts parts of that patch.

bartoldeman commented:
> Not sure I can parse that.
> What does this mean, "both kvm_base and main_pool need to alias
> map into SHM"?

for example:

in KVM we have a locked EMB mapping:
```
GVA -> GPA -> HVA -> SHM
config.dpmi_base-phys_low+xms_base -> xms_base -> &kvm_base[xms_base] -> address from alloc_mapping
```
and outside KVM we also need this mapping:
```
&mem_base[config.dpmi_base-phys_low+xms_base] -> same address from alloc_mapping
```

in a VCPI scenario with $_dpmi=(0) the GVA is then managed by the VCPI client:
```
GVA -> GPA -> HVA -> SHM
(managed by VCPI client) -> xms_base -> &kvm_base[xms_base] -> address from alloc_mapping
```
and the mem_base addresses don't exist.

bartoldeman committed:
mapping: move physical extmem/XMS map to kvm_base

By moving the KVM physical extmem/XMS to kvm_base we will
later be able to free up the highest part of main_pool and
allow all DPMI mappings in the same contiguous chunk below
dpmi_base.

Moreover this approach will be compatible with VCPI.

The cost is that now both kvm_base and main_pool need to alias
map into SHM, for compatibility with native/JIT DPMI.

example:

before:
(GVA -> GPA ->) HVA -> SHM
dpmi_top+align+xms_base -> xms_base -> &mem_base[dpmi_top+align+xms_base] -> SHM

after:
dpmi_top+align+xms_base -> xms_base -> &kvm_base[xms_base] -> SHM
*and*
&mem_base[dpmi_top+align+xms_base] -> SHM

bartoldeman committed:
KVM: allow merging of memory map slots if possible.

This makes sure the low and high parts of kvm_base are merged into one slot
for efficiency.

bartoldeman committed:
mapping: move extmem/XMS DPMI alias below dpmi_base, decrease memsize.

This way it is mapped in the same contiguous linear area of main_pool as
the other linear mappings, regular DPMI memory is at the top, and
memsize = config.dpmi_base + dpmi_mem_size() without any special
alignment needs.

The physical memory for the 'U' mapping (virtual memory at 0x110000) is
now set to the first aligned position beyond memsize.

bartoldeman committed:
mapping: allow register_hardware_ram_virtual with vbase=-1

Up so far we've always had a vbase for virtual hardware, but with
extmem/XMS the vbase is only needed for DPMI.

Allowing vbase=-1 makes it possible to use a minimal main_pool of
just 0x110000 if $_dpmi=(0), since XMS/EMS don't need to be mapped
from main_pool then.

This changes do_get_hardware_ram to always return the pointer to
the struct and set the virtual address optionally, where it was
the opposite before, as we now sometimes only need that pointer but
not the virtual address.

bartoldeman commented:
Rebased and cleaned up history now, so the step are a bit smaller, and register_hardware_ram is never called anymore without _virtual.

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -906,16 +906,18 @@ static dosaddr_t do_get_hardware_ram(unsigned addr, uint32_t size,
-	  *r_hw = hw;
-      return hw->vbase + addr - hw->base;
+    if (hw->base <= addr && addr + size <= hw->base + hw->size) {
+      if (hw->vbase != (dosaddr_t)-1)
```
So you can bypass initializing va and
return -1, and yet have some side-effect
via passed pointer r_hw?
That's a huge programming anti-pattern.

bartoldeman replied:
I've now switched hw and va here, to eliminate that priority issue, and as a bonus it can then be easily used in `get_hardware_uaddr()` as well.

stsp commented on src/arch/linux/mapping/mapping.c:
```diff
@@ -280,8 +280,13 @@ int alias_mapping(int cap, dosaddr_t targ, size_t mapsize, int protect, void *so
+    prot = protect;
+    if (i == KVM_BASE) {
+      /* in KVM targ is physical memory, not equal to virtual memory for high mem */
+      if (targ >= ALIAS_SIZE) continue;
```
This almost suggests you need both
`pa_size` and `va_size` fields in mem_bases.
Not insisting though.

bartoldeman replied:
The only place where this is relevant for alias_mapping() I could find is DPMI mappings to conventional memory (int31/ax=0509). In that case this would be better done by adjusting KVM page tables, so it's already a bit of a special case that should be adjusted (but it's a bit beyond the scope of this PR).

bartoldeman commented:
oops! I realized I pushed the wrong branch yesterday. So some of your comments are out of date, sorry for that.

The clunky interface (anti-pattern) in mapping.c with do_get_hardware_ram I'll improve later though.

stsp commented:
Looks good except the first one.

bartoldeman committed:
extmem: use alias_mapping_pa instead of alias_mapping

Because alias_mapping_pa() calls hwram_update_aliasmap(),
physaddr_to_unixaddr() in int15() now maps to an address in
shared memory, instead of in main_pool (mem_base).

This is now consistent with XMS, except when $_dpmi=(0), where
for extmem physaddr_to_unixaddr() still maps to main_pool,
but XMS maps to SHM.

bartoldeman committed:
extmem: use alias_mapping_pa also for $_dpmi=(0)

This way the extmem mappings are fully consistent with XMS, and
when $_dpmi=(0) the memory at &mem_base[0x110000] is still mapped
to shared memory, but no longer accessed by int15() or anywhere else
(since DOS apps can't access it without DPMI and VCPI support is not
merged yet).

stsp commented:
Ping.
This patch has conflicts.

stsp commented:
This seems to have a lot
of intersections with 517a4c614
Bart, what should we do here?
